Limited implementation for complex matrix multiplication in mmx

AdamCooman opened the pull request:
Using the motto "Better something than nothing" I modified the mmx code to obtain a simple mex file that can perform matrix-matrix multiplication for complex matrices using the same threading machinery as mmx.

I don't use BLAs, so for larger matrices, there are better ways to compute this matrix product, but for small problems, my simple implementation of mmx beats everything, due to the simple threading I borrowed from mmx.
For very small problems, the threads create too much overhead, so I also implemented a version without threads, which is lightning-fast for small problems.

## Implementation

I am not a C or C++ programmer, so there is probably a LOT that can be improved in my implementation. I just went through the original mmx code and added extra variables when they were needed for complex matrices. 

Because it was way too ambitious for me to add the other options using the BLAS, I removed the option to choose the operator and only allow matrix multiplication.

Inside the code, `mmxmultc` checks whether it is dealing with one of the following cases:

+ real matrix times real matrix
+ real matrix times complex matrix
+ complex matrix times real matrix
+ complex matrix times complex matrix

Depending on the problem, it then calls the appropriate function implemented in `matrix_fun_complex.c`

I did not touch the signleton expansion stuff, so it should also work in my implementation. I did not test it thoroughly though.

## Bug

Due to a bug I cannot seem to resolve, the matrix transpose option can crash matlab. 
I therefore throw an error when the use tries to transpose matrices.

## Tests

To test my modified code, I implemented a simple test class using matlab's new test framework. You can run the tests by running the `run_mmxmultc_tests` script. It tests different matrix sizes and tries all combinations of real and complex matrices.

## Future

I am not at all happy with this implementation, but I can't do any better than this. Adding complex support for the other operators and using the BLAs is needed. 

mmx is still by far the fastest way to compute a lot of matrix-matrix multiplications in matlab. We have tried many different techniques using the new automatic expansion in matlab's dot product, but their performance pails in comparison to what mmx can do.

I do research in electronics and control applications and encounter the complex mmx problem all the time. It must be that many other domains can benefit from a complex mmx too.

AdamCooman committed:
implemented blasless complex matrix multiplication mmx

AdamCooman committed:
added threadless version of mmxmultc for small matrices

AdamCooman committed:
added build script for mmxmultc and mmxmultc_nothreads

AdamCooman committed:
added test for mmxmultc

yuvaltassa commented:
Adam, this is fantastic!

I'm tempted to accept it as (and I will if you don't feel like touching it anymore) but don't you think that instead of a complete branch into `complex/`, you could just add the code in the main branch? I understand that currently only a small subset is supported (non-transposed multiplication), but checking and throwing appropriate errors should be easy, no?

It would be so much easier for others to improve/modify/fix your code if the diffs were more informative...

yuvaltassa committed:
Merge pull request #3 from AdamCooman/master

Fix for compilation problems on windows 10

AdamCooman committed:
changed code run in threads to track a real and imaginary part

AdamCooman committed:
added imaginary part for inputs and outputs

AdamCooman committed:
added functions for complex matrix multiplication

AdamCooman committed:
replaced mulMatMat by its several complex cases

AdamCooman committed:
added real and complex cholesky functions

AdamCooman committed:
added real and complex solve functions

AdamCooman committed:
added real and complex square functions

yuvaltassa commented:
Could you please switch your editor from tabs to 3 spaces like the rest of the code?
[I haven't had time to patch and test yet]

AdamCooman commented:
I'll make sure the indents are correct.

I'm working on adding the complex stuff to the main file, it will take a while though, as I have to make sure I don't introduce errors in the other guys.

AdamCooman commented:
ow, I just noticed that my new commits are already added to my earier PR.
The code I commited there does not work at all, I'm working on it. I just pushed that stuff to switch to a computer that has a compiler.

Dev-iL commented:
@AdamCooman Yeah, I had this problem myself in the past. You should rebase these last few commits onto a new feature branch.

AdamCooman committed:
moved printf before call to find crashed easier

AdamCooman committed:
fixed bug in solveRC call

AdamCooman committed:
Throw an error when the complex version is not implemented

AdamCooman committed:
fix bug in accessing the complex part of B

AdamCooman committed:
fixed bug where wrong function was called in realxreal

AdamCooman committed:
fixed some bugs related to calling backslash

AdamCooman committed:
function now calls naive implementation for complex multiplication, even when blas are used for real

AdamCooman committed:
fixed wrong function calls in square

AdamCooman committed:
added test

AdamCooman committed:
replaced tabs by triple spaces

AdamCooman committed:
ensured that chol input is positive definite

AdamCooman committed:
enabled backslash test on square matrices

AdamCooman committed:
deleted old complex code

AdamCooman commented:
Ok, I added the complex stuff to the mmx code now and tested it. It works on my computer. (I added a new test too, which uses the new matlab testCase objects)

When mmx is compiled with the `USE_BLAS` option, it uses BLAs for the real operators. when a complex matrix is provided, it falls back onto the stupid implementation of complex multiplication. When `USE_BLAS` is disabled, it always uses the stupid implementations.

On my computer, the `build_mmx` function was not able to include the blas library of matlab. I had to copy them to the mmx folder to make it work. I have no idea why.

AdamCooman commented:
https://mathworks.com/help/matlab/matlab_external/matlab-support-for-interleaved-complex.html

> As of MATLAB Version 9.4 (R2018a), MATLAB uses an interleaved storage representation for C language MEX files, where the real and imaginary parts are stored together.

This will break my code in 2018a+, but it should make it a lot easier to add support for complex BLAs in the future.
